libs/tlmatcher/src/main/java/org/codice/util/tlmatcher/TimeLimitedMatcher.java

```diff
@@ -0,0 +1,216 @@
+   * @return The (possibly empty) subsequence captured by the group during the previous match, or
+   *     null if the group failed to match part of the input
+   * @see Matcher#group(int)
+   */
+  public String group(int group) {
+    return matcher.group(group);
+  }
+
+  /**
+   * Returns the input subsequence captured by the given named-capturing group during the previous
+   * match operation.
+   *
+   * @param name The name of a named-capturing group in this matcher's pattern
+   * @return The (possibly empty) subsequence captured by the named group during the previous match,
+   *     or null if the group failed to match part of the input
+   * @see Matcher#group(String)
+   */
+  public String group(String name) {
+    return matcher.group(name);
+  }
+
+  /**
+   * Returns the number of capturing groups in this matcher's pattern.
+   *
+   * @return The number of capturing groups in this matcher's pattern
+   * @see Matcher#groupCount()
+   */
+  public int groupCount() {
+    return matcher.groupCount();
+  }
+
+  /**
+   * True if the entire region matched against the pattern.
+   *
+   * @return true if the entire region matched against the pattern.
+   * @see Matcher#matches()
+   */
+  public boolean matches() {
+    return matches;
+  }
+
+  public String toString() {
+    return String.format("TimeLimitedMatcher of [%s]", matcher.toString());
+  }
+
+  private static long calculateEndTime(Duration duration) {
+    return System.currentTimeMillis() + duration.toMillis();
+  }
+
+  /** Time-limited character sequence. */
+  static class TLCharSequence implements CharSequence {
+    private final CharSequence delegate;
+    private final long endTime;
+    private boolean isRestricted;
+
+    /**
+     * Default constructor.
+     *
+     * @param delegate The CharSequence to wrap.
+     * @param endTime When to abandon the regex.
+     * @param isRestricted if processing should be time-limited or not
+     */
+    private TLCharSequence(CharSequence delegate, long endTime, boolean isRestricted) {
+      this.delegate = delegate;
+      this.endTime = endTime;
+      this.isRestricted = isRestricted;
+    }
+
+    private void unrestrict() {
+      isRestricted = false;
+    }
+
+    @Override
+    public char charAt(int index) {
+      // This is an unavoidable slowdown
+      if (isRestricted && System.currentTimeMillis() > endTime) {
+        throw new TimeoutException("RegEx timed out");
+      }
+      return delegate.charAt(index);
+    }
+
+    @Override
+    public int length() {
+      return delegate.length();
+    }
+
+    @Override
+    public CharSequence subSequence(int start, int end) {
+      // Ensure that any subsequence generated during a regular expression
+      // operation is still going to fail on time, if this CharSequence is still time-restricted
+      return new TLCharSequence(delegate.subSequence(start, end), endTime, isRestricted);
```

Comment:
❓ Shouldn't they all share the same isRestricted flag such that when it is turned off above, they are all off?